### PR TITLE
feat(performance): identify and select the GPU behind the live readings

### LINKED ISSUE #1806: feat(performance): identify and select the GPU behind the live readings
> *This was generated by AI during triage, corrected after maintainer clarification, and rescoped after the Performance / System Specifications redesign shipped (#1942, #1943).*

## Parent

- Epic: #1793
- Initial tab shell: #1796
- Redesign that changed the surfaces this issue targets: #1942, #1943

## Agent Brief

**Category:** enhancement
**Summary:** Make the GPU whose live readings are on screen identifiable and selectable on the Performance Screen. This is not a standalone GPU screen or navigation route.

**Why the scope moved from System Specifications to Performance:**
When this issue was filed, both surfaces were tabs inside one Grouped Dashboard destination and the specifications tab reused the Hardware Dashboard GPU card, so adapter inventory and live readings sat together. ADR 0014 then split them deliberately, and ADR 0015 promoted both to peer sidebar destinations:

- The System Specifications Screen is a flat sheet of static facts. It already lists **every** detected adapter with name, vendor, memory, and dedicated memory. It intentionally carries no live readings, so a "Selected GPU" has nothing to select there.
- The Performance Screen owns every live GPU reading (usage, temperature, VRAM, fan) - and it is the surface that currently shows those readings **without saying which adapter they came from and without any way to change the adapter**.

So the remaining gap is on Performance, and the acceptance criteria below are stated against that screen.

**Current behavior:**
The Performance Instrument Strip and Compact strip resolve one effective GPU and render its usage, temperature, VRAM, and fan values with no adapter identity shown. The only GPU selection control in the app lives in the classic Hardware Dashboard (`GPUInfo`), so a user on grouped navigation cannot reach the second adapter's readings at all.

**Desired behavior:**
The Performance Screen states which adapter its GPU readings belong to and, on a multi-adapter machine, lets the user choose. An explicit choice is honored even when that adapter reports nothing, in which case the screen says so rather than borrowing another adapter's numbers.

**Key interfaces:**
- `GraphicInfo.id` and the per-GPU live-value maps (`gpuUsageHistoriesAtom`, `gpuTempMapAtom`, `gpuFanSpeedMapAtom`, `gpuDedicatedMemoryKbMapAtom`) define adapter identity.
- `selectedGpuIdAtom` is UI-local explicit user intent, persisted by `useSelectedGpuPersistence`.
- `getEffectiveGpuId` in `InstrumentStrip.tsx` is the single join point shared by both Performance views.

**Acceptance criteria:**

- [ ] The Performance GPU instrument names the adapter its readings belong to, on single- and multi-adapter machines.
- [ ] Every detected adapter is reachable from the Performance Screen, and the selected one is visually unambiguous.
- [ ] Selecting an adapter that reports no live values keeps that selection and shows an explicit unavailable state, instead of falling back to another adapter's readings.
- [ ] The unavailable state does not hide the rest of the adapter's information.
- [ ] "Not measured yet" is not presented as "unavailable" before the first sample arrives.
- [ ] The Compact strip identifies the adapter behind its GPU row.
- [ ] The selection survives Performance view changes, navigation between destinations, navigation-layout changes, and restarts while the adapter is still detected.
- [ ] No GPU navigation route or standalone GPU screen is introduced.
- [ ] Focused multi-GPU tests cover selection, cross-adapter attribution, and the unavailable state.
- [ ] Rendered desktop and compact layouts are inspected.

**Already delivered (#1942 - do not redo):**

- Live values are joined to the effective adapter by stable `GraphicInfo.id`; no metric is borrowed across adapters (`getEffectiveGpuId`, plus the `Performance.test.tsx` cases *"shows the temperature for the GPU selected by the usage history"* and *"uses one effective GPU for history and temperature fallbacks"*).
- The System Specifications sheet represents every detected adapter as inventory with `#n` badges, separated from live readings.
- A missing temperature, VRAM, or fan reading is omitted rather than rendered as zero.

**Out of scope:**
- Changing GPU collection providers or archive semantics.
- Redesigning the GPU view inside the Insights Screen.
- Adding selection or live readings to the System Specifications sheet (ADR 0014 keeps it static).
- Changing the classic Hardware Dashboard GPU card.

### PR DESCRIPTION
**Pull request by shm11C3**
Closes #1806.

## Problem

ADR 0014 moved every live GPU reading onto the Performance Screen. It never said which adapter produced them, and grouped navigation had no GPU selection control at all — `setSelectedGpuId` was called in exactly one place, the classic Hardware Dashboard `GPUInfo` card. A user with a discrete and an integrated GPU could neither tell which adapter the numbers described nor reach the other one.

The issue was filed while both surfaces were tabs in one Grouped Dashboard destination and assumed the specifications tab carried live readings. It was rescoped to the Performance Screen before implementation ([comment](https://github.com/shm11C3/HardwareVisualizer/issues/1806#issuecomment-5308182464)); the criteria #1942 already met were moved to an `Already delivered` section rather than redone.

## Change

- **The GPU instrument names its adapter.** One adapter: the name alone. More than one: a named switcher writing `selectedGpuIdAtom`, which `useSelectedGpuPersistence` already persists.
- **Explicit selection beats data availability.** The resolver used to drop a selection the moment its adapter left the usage map, silently substituting another adapter's numbers under the same label. It now honors the selection while the adapter is still detected and reports nothing when it reports nothing. Only a selection pointing at an adapter that no longer exists falls back (DP-06).
- **"Not measured yet" is not "unavailable."** The no-readings note is withheld until some adapter has reported, and it is additive — it never takes the place of what the adapter did report (DP-02).
- **Every live map counts.** Usage, temperature, fan speed, and dedicated memory are populated independently, so all four are consulted before calling an adapter silent. They travel together as `GpuLiveMaps`.
- **Compact names its adapter** in the footer; the row tracks are sized for the mini monitor's small corner window and cannot hold a device name. Compact follows the choice made in Panels and offers no selection of its own.
- **Labels drop only what cannot distinguish adapters** — trademark marks, a leading vendor word, and a leading run of words every adapter repeats. That last rule is what keeps "GeForce RTX 4080" and "GeForce RTX 4060" from both rendering as `GeForce RTX 40…` in a three-column strip. If shortening would make two labels identical, all fall back to full names. The full name stays as the accessible name and tooltip.

The selector is a labelled group of toggle buttons rather than a tablist: there is no tabpanel, roving tabindex, or arrow-key contract to back that role up.

## Decision record

New [ADR 0016](docs/adr/0016-gpu-attribution-on-the-performance-screen.md), refining ADR 0014/0015. `CONTEXT.md` gains **Selected GPU**. The System Specifications sheet is untouched — ADR 0014 keeps it static facts, so it has no reading to attribute and no selected state.

## Verification

- 510 unit tests pass, including 21 new ones across `gpuIdentity.test.ts` and `Performance.test.tsx`: silent-adapter selection is kept and non-borrowing is asserted positively *and* negatively; the removed-adapter fallback, the pre-first-sample silence, fan-only and VRAM-only adapters, label collisions, and the cross-view carry are each pinned.
- 20 e2e pass, including a new multi-GPU case: both adapters reachable, switching moves the VRAM reading from `4.0/8 GB` to `1.0/2 GB`, the choice survives navigating away to System Specifications and back, and Compact reports the same adapter. Captures `performance-gpu-selector-desktop` and `performance-gpu-selector-compact`.
- Rendered layouts inspected at 1280 / 820 / 520 px. `tsc --noEmit`, `biome check`, and `check:agent-guidance` clean.

## Review note

A Fable review of the first commit found that the honesty check was itself dishonest — it consulted only two of the four live maps, so a fan-only adapter was declared silent *and* had its real fan/VRAM readings hidden by the note that replaced the substats. That is the second commit, along with the raw-LUID naming path, the ARIA role, and the hard-coded footer separator.

## Known limitation (not introduced here)

The live maps are append-only in `useHardwareEventListener`, so an adapter that reports and then goes silent keeps its last value on screen. The unavailable state only covers an adapter that has never reported. Recorded in ADR 0016's consequences; distinguishing stale from current needs per-sample presence tracking in the event listener and is filed separately as #1945. A second review finding, #1946, covers a Selected GPU being discarded after one launch without that adapter.

## Summary by CodeRabbit

- **New Features**
  - Added GPU adapter selection to Performance, Monitor, Dashboard, and Compact views.
  - Preserves the selected adapter across navigation and application restarts.
  - Displays adapter names with accessible labels, tooltips, and localized text.
  - Indicates when an adapter is detected but has no live readings.
  - Keeps GPU usage, temperature, fan, and VRAM readings aligned with the selected adapter.

- **Bug Fixes**
  - Improved GPU identification when inventory and live monitoring identifiers differ.
  - Prevented ambiguous adapter names from showing incorrect readings.

- **Documentation**
  - Documented GPU attribution and adapter-selection behavior.

### CODE REVIEW

**Comment by shm11C3** [comment]
## Review round 2 — pushed `504a7303`

### Codex P1: confirmed and fixed

I verified the claim in the Rust sources and it is correct, and broader than the comment stated — the inventory and the monitor stream use different id namespaces on **every** platform, not just Windows NVIDIA:

| platform | inventory `GraphicInfo.id` | live `GpuSample.gpu_id` |
|---|---|---|
| Windows NVIDIA | raw NVAPI id (`nvapi_provider.rs:193-203`) | `nvapi:<id>` (`platform/windows/gpu.rs:138`) |
| Windows other | DXGI `device_id` (`directx.rs:76,120`) | `pci:<b>:<d>:<f>` / `pdh:<name>` (`gpu.rs:250,282`) |
| macOS | `0x<registry_id>` / `spdisplay-<i>` (`macos/gpu_info.rs:29,160`) | `iokit:<name>` (`platform/macos/gpu.rs:51`) |
| Linux | `card<n>` (`platform/linux/gpu.rs:78,92`) | PCI BDF (`gpu.rs:195-200`) |

So the union I wrote would have rendered one physical GPU as two adapters on real hardware, one of which reports nothing and would then have been declared silent by the very feature this PR adds. Fixed by removing the join entirely: the adapter list is built from the live side alone, named by the `gpuName` every sample already carries (new `gpuNamesAtom`). The inventory keeps its own surface, the System Specifications sheet.

Two further consequences of the same fact, both fixed here:

- **The VRAM total was already broken** by an id lookup into `hardwareInfo.gpus` (shipped in #1942). It now joins on the name, the same join `GPUInfo` uses for temperature.
- **`useSelectedGpuPersistence` validated the stored id against the inventory**, which it can never match — so with the selector writing live ids, the restart-persistence criterion would have silently failed. It now restores the stored id as-is, which also fixes the DP-06 problem in #1946 (intent overwritten permanently after one launch without the adapter).

### The fixture was the reason this passed

`GPU_FIXTURES` used one id for both the inventory and the sample payload, which no real platform does. It now carries distinct `id` and `liveId`, and the e2e asserts exactly one control per physical adapter — so a cross-namespace join fails in CI rather than on a users machine.

### CodeRabbit

- **Fallback order in ADR 0016** — fixed, now states usage → temperature → fan speed → dedicated memory.
- **Full name for non-button labels** — the single-adapter label now carries an `sr-only` full name alongside the shortened visible text, rather than relying on `title`. The Compact footer is a plain text line rather than a control; its shortened label is the deliberate choice (the mini monitors row tracks cannot hold a full device name, which is why the name is in the footer at all), and the full name is available on the Panels view.

### Filed separately

- #1947 — the classic dashboard GPU selector writes an inventory id into `selectedGpuIdAtom`, so switching to `#2` silently keeps showing `#1`. Pre-existing; ADR 0016 records it as a non-goal.
- #1945 — live maps are append-only, so a GPU that stops reporting keeps its last value.
- #1946 — largely fixed here; the persistence hook no longer discards intent.

510 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
Both round-2 notes addressed in the commit above.

- **Full name in non-button labels** — the Compact footer now carries the unshortened name for assistive technology and hover alongside the shortened visible text. (The single-adapter paragraph was already fixed in `504a7303`.) Footer entries gained a small `{ id, text, fullText }` shape so the not-yet-collected entries the strip is designed to grow into can do the same.
- **Event-listener regression test** — added to `useHardwareEventListener.test.ts`, asserting the id-to-name mapping for two adapters with distinct live ids, one of which reports nothing but its own identity. That last case is the one the unavailable state depends on, so it is worth pinning at the fan-out level rather than only in the component test.

511 unit tests and 20 e2e pass.

File: src/features/hardware/performance/components/GpuAdapterSelector.tsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      <p
+        className={cn(
+          "min-w-0 truncate text-muted-foreground text-xs",
+          className,
+        )}
+        title={adapter.name}
+        data-testid="performance-gpu-adapter"
+      >
+        {adapter.label}
+      </p>
```
[the remark this answers is not in the document]
Comment (shm11C3) [reply src/features/hardware/performance/components/GpuAdapterSelector.tsx:53]: Both sites now expose the full name; this thread was re-anchored onto the newer commit rather than re-found.

- `GpuAdapterSelector.tsx:44-45` — the single-adapter paragraph renders `adapter.label` as `aria-hidden` and `adapter.name` in an `sr-only` span, so the accessible name is the full one rather than relying on `title` (fixed in `504a7303`).
- `CompactStrip.tsx:224-232, 252-260` — `footerItems` is now `{ id, text, fullText }`; the GPU entry keeps `activeGpuAdapter.name` in `fullText`, which is rendered `sr-only` and set as the `title` (fixed in `ea798637`).

The names are not separately localized because an adapter name is a device string the platform reports, not a translatable phrase; the surrounding sentence goes through `pages.performance.compactGpuAdapter` in en/ja/ru.

**Comment by shm11C3** [comment]
@coderabbitai review

**Comment by shm11C3** [comment]
## Review round 3 — pushed `a1f8b68a`

All four findings confirmed in the code and fixed. The P1 was mine, introduced in `504a7303`.

### P1 — the persistence hook erased the preference it had just restored

Confirmed by reading the effect order. `useSelectedGpuPersistence` flipped `hydratedRef` **inside** the hydration effect, so the write-back effect — running later in the *same* commit, with `selectedGpuId` still at its pre-hydration value — saw `hydratedRef.current === true` and called `setStoredId(null)`. The stored id was wiped on the first render after the store loaded, every launch.

The gate is now `useState`, so the write-back effect is skipped until a later render, by which point the atom holds the restored value and the equality check short-circuits. New `useSelectedGpuPersistence.test.tsx` pins this directly (`does not write the pre-hydration value back over the restored one`) plus restore, explicit change, absent-adapter, and store-pending.

### P2 — two identical cards were indistinguishable

Correct, and my own test asserted the broken behaviour. Both controls fell back to the same full name *and* the same `aria-label`, so the selector could not say which card was being selected. Duplicated names now get an ordinal (`GeForce RTX 4090 #1` / `#2`), applied **before** the full-name fallback so the label stays short enough for the card. The accessible name follows the same rule, so the ordinal is announced exactly when the raw name cannot tell them apart.

### P2 — VRAM total attributed through a non-unique name

Correct. `GpuAdapter` now carries `isNameAmbiguous`, and the inventory join is refused when it is set: the denominator is dropped (`VRAM 4.0 GB`) rather than taken from whichever identical card `.find()` happened to hit first. Showing a live reading against the wrong card capacity is worse than showing no capacity.

### P2 — partial samples erased adapter identities

Correct. `setGpuNames` replaced the whole map, so a provider omitting one adapter from a single tick deleted the only record that it exists, and `getEffectiveGpuId` jumped to another GPU on a transient hiccup. A name is identity, not a reading, so the map now merges. Test added alongside the fan-out test.

---

The CodeRabbit event-listener-test thread was already addressed in `ea798637`; the accessible-name thread was addressed in `504a7303` and `ea798637` and I replied inline with the line references.

518 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
## Review round 4 — pushed `45745470`

Two new findings fixed, two stale threads, one declined with reasoning.

### Fixed: VRAM total, ambiguity on the inventory side

Correct, and a real gap in my previous fix. `isNameAmbiguous` counts **live** adapters only, so an inventory holding two identically named cards while only one of them reports passed the guard, and `.find()` attached whichever capacity came first — exactly the case ADR 0016 permits (an adapter present in the inventory but not in the stream). The join now requires the name to be unique on both sides. Two tests added, one per direction.

### Fixed: Compact lost the ordinal in its accessible text

Correct. `fullText` used `activeGpuAdapter.name` — the very string two identical cards share — while the ordinal-bearing visible label was `aria-hidden`. Assistive technology got a name that could not identify the card. It now uses the label when `isNameAmbiguous` is set.

### Stale (re-anchored, already fixed)

- **P1 hydration overwriting the stored GPU** — the comment body still describes `hydratedRef` and links to `504a7303`. Fixed in `a1f8b68a`: the gate is `useState`, so the write-back effect is skipped until a post-hydration render. Pinned by `useSelectedGpuPersistence.test.tsx` → *does not write the pre-hydration value back over the restored one*.
- **P2 identical full names** — links to `ea798637`. Fixed in `a1f8b68a`: duplicated names get an ordinal, applied before the full-name fallback, and the accessible name follows the same rule. Pinned by *adds an ordinal when the platform reports the same name twice*.

### Declined: routing `selectedGpuId` through typed Rust settings commands

`.agents/rules/settings.md` draws the line as:

> - **Application Preference**: a user-facing choice expected to survive restart as app configuration. Persist it in `settings.json` through typed Rust IPC […]
> - **UI-local State**: a resettable selection, cache, or view state that is not an explicit configuration. It may use Tauri Store.

A selected GPU is a resettable selection, not app configuration — it does not appear on the Settings screen and resetting it changes nothing but which device the screen describes. #1806 classifies it the same way ("`selectedGpuId` is UI-local explicit user intent"), and its sibling `selectedStorageDeviceId` uses the same `useTauriStore` path under ADR 0008. That rule file's `scope` also does not cover this hook.

This PR did not introduce the storage choice either; it predates the branch. Migrating it would be a settings-schema change across Rust and the generated bindings, which is out of scope here and, on the above classification, not warranted. Happy to file it if the maintainer reads the boundary differently.

520 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
## Review round 5 — pushed `9d7dc71b`

One genuinely new finding this round; the other four threads are re-anchors of ones already answered (their bodies still cite `hydratedRef` and link to `504a7303` / `ea798637`).

### Fixed: the shared selection reached Classic in the wrong namespace

Confirmed, and this PR is what made it reachable. `selectedGpuIdAtom` is shared, and the Performance selector writes a **live** id into it. `GPUInfo` resolved that atom with `gpus.find(g => g.id === selectedGpuId) ?? gpus[0]` — an id match that can never succeed across the two namespaces — so it fell back to the first inventory entry and rendered **that** adapter's name, badge, and highlight beside the **selected** adapter's usage, source, and VRAM, which resolve through the live maps.

Before this PR the two happened to agree, because the only live id in the atom was the event listener's auto-selection of the first adapter, and `gpus[0]` is also the first inventory entry. Making the second adapter selectable broke that coincidence.

Both directions now go through the name, and only while the name picks out exactly one entry on each side:

- `findInventoryGpu` reads the shared selection back into an inventory entry.
- `toLiveGpuId` lets the classic selector write the selection in the namespace every reading is keyed by, so Performance honours it too.

`gpuIdentity` moved from `performance/` up to the hardware feature root, since both surfaces now use it. 7 new unit tests cover both helpers including the ambiguous-name refusals.

**The classic e2e was asserting only the pressed state**, which is why it passed through all of this — the same gap I noted when filing #1947. It now asserts the rendered readings actually change, and I verified it **fails** without this commit (reverted the fix locally, test red; restored, green).

### Still standing from earlier rounds

- **P1 hydration / P2 identical names** — fixed in `a1f8b68a`, replied inline with the test names.
- **Compact ordinal / inventory-side VRAM ambiguity** — fixed in `45745470`.
- **Typed settings command for `selectedGpuId`** — declined with reasoning in the previous comment: `.agents/rules/settings.md` puts a resettable selection in the UI-local category, #1806 classifies it that way, `selectedStorageDeviceId` uses the same path under ADR 0008, and that rule's `scope` does not cover this hook. Happy to file it if the maintainer reads the boundary differently.

527 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
## Review round 6 — pushed `c0c77aaa`

Both new findings were fair, and the first one was a gap in my own acceptance criterion rather than a corner case.

### Fixed: Monitor had no attribution at all

Correct. Monitor mounts only `UsageGraphPanel`, labelled with the generic `GPU` series name, and the Instrument Strip that carries the adapter identity is not mounted there. A multi-GPU user in Monitor could neither see nor change which adapter the yellow line described — while the PR claims the Performance Screen names the adapter behind its GPU readings.

The selector now sits in the Monitor toolbar whenever GPU is among the graph's display targets (it is pointless when the user has turned the GPU series off). Verified in the browser and asserted in the e2e, with a `performance-gpu-selector-monitor` capture.

### Fixed: twin cards fell back to a claimed identity

Correct, and it put back the exact mislabelling the join was added to prevent, narrowed to two identically named cards. `findInventoryGpu` refused the ambiguous join, but the caller's `?? gpus[0]` then claimed the first entry anyway.

Split by case:

- **Nothing selected** → the first entry is still the honest default.
- **A selection that cannot be resolved** → no identity is claimed. No badge is highlighted and no name-matched temperature is attributed, rather than labelling one adapter's readings with another's name.

I did **not** take the other option of pairing the two sides by position among same-named entries. It would look plausible and is a guess — the inventory's enumeration order and the stream's are different enumerations, so a wrong pairing would produce silent mislabelling of exactly the kind this whole thread is about. There is a test asserting the refusal (`refuses rather than pairing by position when both sides have twins").

### Also in this commit

The three GPU surfaces (Instrument Strip, Compact, Monitor) now read their adapter list, effective adapter, and reporting state from one `useGpuAdapters` hook instead of each re-deriving them from six atoms. That is what made the Monitor fix a three-line change, and it removes the possibility of two views disagreeing about which adapter they describe.

528 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
## Review round 7 — pushed `3cd3301e`

Both findings were regressions from the Monitor selector I added last round. Both are fixed, and both now have a test that **fails without its fix** — I verified each by reverting locally.

### P1 — Monitor named one adapter and graphed another

Confirmed. The selector resolves through `getEffectiveGpuId`, which honours an explicit selection even when that adapter reports no usage. `UsageGraphPanel` reads `graphicUsageHistoryAtom`, which fell back to the first history whenever the selected key was missing. Selecting an iGPU that reports a temperature but no usage therefore put the iGPU's name in the toolbar and the discrete card's line on the graph.

All three derived atoms (`graphicUsageHistoryAtom`, `gpuUsageSourceAtom`, `gpuDedicatedMemoryKbAtom`) now resolve through the same rule the selectors use, via a shared `effectiveGpuIdAtom`. A valid selection with no readings yields nothing rather than another adapter's numbers.

This also removes the same borrowing from the **classic Usage screen and dashboard**, which read those atoms too — so the fix is wider than Monitor. New `store/chart.test.ts` covers the four cases (no usage of its own, resolves, selection gone, before first sample).

### P2 — the subscription sat on the Performance parent

Confirmed, and it is the ADR 0010 rendering-cost rule this repo has a documented incident about. `useGpuAdapters` subscribes to five atoms the event listener rewrites every sample; calling it from `Performance` meant Panels rerendered its toolbar and every panel at 1 Hz, instead of the update staying inside the GPU instrument.

Moved into a `MonitorGpuSelector` child mounted only in Monitor, which also owns the display-target gate. New test *keeps a GPU tick out of the panels that do not show GPU data* asserts the process table and the CPU/RAM charts do not rerender on a GPU tick; it fails with the subscription restored to the parent.

533 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
## Review round 8 — pushed `74bb1801`

Two fixed, one deferred to an existing issue with reasoning.

### Fixed: Monitor did not report a silent selection

Correct. Monitor shows only the graph, so an adapter selected while it reports nothing gave a named selection above a blank series with no explanation — a blank line reads as idle, which is the DP-02 failure the unavailable state exists to prevent, and Panels labels the same state. `MonitorGpuSelector` now renders the existing localized message beside its selector; it already had `hasNoReadings` from the shared hook and simply was not using it.

### Fixed: Classic could commit an id that addresses no readings

Correct. The classic card is reachable before the first sample, when `gpuNames` is still empty, so `toLiveGpuId` fell through to the inventory id. `findInventoryGpu` then matched it by id and highlighted the requested adapter, while `getEffectiveGpuId` rejected it and every graph kept showing the first live adapter — and persistence carried that mismatch across restarts.

The card now adopts the live id as soon as the stream names that adapter. I kept the fallback rather than deferring the selection, because refusing to record a click until a sample arrives would silently drop a deliberate action; reconciling after the fact keeps the click and converges within one tick.

### Deferred: retiring names after sustained removal → #1945

Accurate, and it is the same defect as the one already filed: the live maps have no deletion path, so an adapter that is named once and unplugged stays in the selector for the session, exactly as a stopped adapter's last usage value stays on the graph.

I have deliberately not added an ad-hoc deletion here. Retiring on first absence would reintroduce the flapping that the merge in `ea798637` was added to prevent — one skipped sample would drop the adapter and move the user's selection. Doing it correctly needs per-sample presence tracking that distinguishes a hiccup from an established removal, which is a change to the event listener's update model and touches all five maps, not just the one this PR added.

I have [scoped that into #1945](https://github.com/shm11C3/HardwareVisualizer/issues/1945#issuecomment-5308749232), and ADR 0016's consequences now name the limitation, the name map included.

534 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
## Review round 9 — pushed `fd29ce8e`

One fixed, one filed as a backend issue.

### Fixed: Compact did not distinguish silent from not-yet-measured

Correct, and it was the one surface of the three still missing it. Compact showed a bare dash, which reads identically to startup before the first sample — exactly the distinction Panels and Monitor make. The footer, which already names the adapter, now carries the same localized message. Test added asserting both the message and that the row still belongs to the *selected* adapter rather than the reporting one.

### Filed: `pdh:<name>` collapses two same-model Intel adapters → #1948

Confirmed in the source. `core/src/platform/windows/gpu.rs:281` keys Intel samples by name while the loop already holds `gpu.luid_high` / `gpu.luid_low` — it queries PDH with them a few lines earlier. Two adapters sharing a DXGI name therefore produce one id, the second overwrites the first in every frontend map, and the adapter is not merely unselectable but absent. The other Windows paths do not have this (`nvapi:<id>`, `pci:<b>:<d>:<f>`).

I have not fixed it here. It is a change to a collection provider, which #1806 and ADR 0016 both exclude, and the id format is not a free change: `selectedGpuId` is persisted, so affected users' stored selection would be invalidated once (degrading to a reset, since `getEffectiveGpuId` falls back — but that should be a deliberate call), and the archive's keying needs confirming. That reasoning and the suggested `pdh:<luid_high>:<luid_low>` shape are in #1948.

Worth noting the frontend behaviour is correct given the input: it renders one adapter because it receives one id. The defect is upstream of everything this PR touches.

535 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
## Review round 10 — pushed `ce6c84e9`

Both findings were real and both were consequences of my own earlier changes.

### Fixed: stored inventory ids from shipped versions never migrated

This is the more serious of the two, and it is an upgrade-path bug I introduced. Shipped versions' classic card wrote `GraphicInfo.id` into the `selectedGpuId` store key. The reconciliation I added in round 8 lived inside `GPUInfo` — which **grouped navigation never mounts**, and grouped navigation is the default. So an upgrading user who had selected a non-first GPU in Classic would have had that stored choice silently ignored for the life of the installation: `getEffectiveGpuId` cannot match it, nothing rewrites it, and the write-back effect sees `selectedGpuId === storedId` so it never even self-corrects.

The translation moved to `useSelectedGpuPersistence`, which `App.tsx` mounts for the whole app. A stored **live** id that is merely absent this session is still left untouched — that distinction is the point of the earlier persistence fix, so there is a test for each direction.

### Fixed: an all-null sample never earned the unavailable message

Correct. `hasNoLiveGpuReadings` required a value in one of the four maps as proof that sampling had happened. A machine whose only GPU reports its name and nothing else — the Intel adapter whose PDH usage query fails, which is exactly #1948's hardware — leaves all four maps empty *permanently*, so `anyAdapterReported` stayed false and all three surfaces showed blanks with no explanation forever.

A detected adapter is now itself proof that a sample arrived. The startup case is unchanged: with nothing detected yet, nothing is claimed.

---

This closes out the review comments other than the ones I have already answered as stale re-anchors or deferred with reasoning (#1945 for retiring names, #1948 for the Intel id collision, and the typed-settings classification). 539 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
## Review round 11 — pushed `69ccd516`

One finding, confirmed and fixed. (The earlier `test-e2e-web` failure on this PR was infrastructure — a 429 fetching the safe-chain installer before any test ran; the rerun passed.)

### Fixed: the migration depended on an inventory fetch nothing performed

Correct. The app-level migration added in round 10 reads `hardwareInfo.gpus`, but the only Performance component that calls the one-shot `init()` is `InstrumentStrip` — and a restart into a persisted **Monitor or Compact** view never mounts it. The stored legacy id then stayed untranslated and exactly the views the migration was meant to serve fell back to the first live GPU.

The migration now performs the fetch itself, gated so it stays cheap:

- It **waits for the first sample** before deciding anything — before that, every id looks unresolved and fetching would be guesswork (and, as a test proved, it would fire on every startup during the pre-sample window).
- After the first sample, a stored id that resolves live costs **no** fetch; only an id the stream does not know triggers one.

Tests cover both directions (`fetches the inventory itself when a stored id cannot be resolved` / `does not fetch the inventory while the stored id resolves live`), and ADR 0016 records the boundary.

541 unit tests and 20 e2e pass.

**Comment by shm11C3** [comment]
@coderabbitai review

File: src/features/hardware/gpuIdentity.test.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+  it("keeps full names when trimming the shared prefix would collide", () => {
+    // Two identical cards: shortening cannot tell them apart, so it must not
+    // pretend to by showing the same short label twice.
+    const adapters = listGpuAdapters(
+      names(
+        ["gpu-1", "NVIDIA GeForce RTX 4090"],
+        ["gpu-2", "NVIDIA GeForce RTX 4090"],
+      ),
+      live({ usageHistories: { "gpu-1": [20], "gpu-2": [5] } }),
```
[the remark this answers is not in the document]
Comment (shm11C3) [reply src/features/hardware/gpuIdentity.test.ts:119]: The anchored test covers two *distinct* full names — "NVIDIA RTX A2000" vs "AMD RTX A2000" — where falling back to the full name is the fix, not the bug. The identical-name case this comment describes is handled and tested directly below it: `adds an ordinal when the platform reports the same name twice` asserts `GeForce RTX 4090 #1` / `#2`, implemented in `a1f8b68a` (`withOrdinals` in `gpuIdentity.ts`), and `GpuAdapterSelector` uses the ordinal-bearing label as the accessible name whenever `isNameAmbiguous` is set.

**Comment by shm11C3** [comment]
@coderabbitai resolve

**Comment by shm11C3** [comment]
@coderabbitai review
